api/src/test/java/org/openmrs/obs/AbstractHandlerTest.java

```diff
@@ -0,0 +1,177 @@
+		
+		File previousFile = null;
+		File currentFile = null;
+		
+		// Mocked methods
+		mockStatic(Context.class);
+		when(Context.getAdministrationService()).thenReturn(administrationService);
+		when(administrationService.getGlobalProperty(any())).thenReturn(complexObsTestFolder.newFolder().getAbsolutePath());
+		
+		// Construct 102 Obs and save them, alternating between 'A' and 'B'
+		for (int i = 0; i <= 101; i++) {
+			String currentData;
+			String previousData;
+			
+			if (i % 2 == 0) {
+				currentData = content1;
+				previousData = content2;
+			} else {
+				currentData = content2;
+				previousData = content1;
+			}
+			
+			ComplexData complexData = new ComplexData(FILENAME, currentData);
+			
+			Obs obs = new Obs();
+			obs.setComplexData(complexData);
+			
+			currentFile = handler.getOutputFileToWrite(obs);
+			
+			// Write the file
+			try (BufferedWriter fout = new BufferedWriter(
+			        new OutputStreamWriter(new FileOutputStream(currentFile), StandardCharsets.UTF_8))) {
+				fout.write(currentData);
+			}
+			
+			// Read the file to check if it wrote properly
+			try (BufferedReader fin = new BufferedReader(
+			        new InputStreamReader(new FileInputStream(currentFile), StandardCharsets.UTF_8))) {
+				String readData = fin.readLine();
+				assertEquals(readData, currentData);
+			}
+			
+			// Read currently previous file if it is not the first, then check if it wasn't overwritten
+			if (i > 0) {
+				try (BufferedReader fin = new BufferedReader(
+				        new InputStreamReader(new FileInputStream(previousFile), StandardCharsets.UTF_8))) {
+					String readData = fin.readLine();
+					assertEquals(readData, previousData);
+				}
+			}
+			
+			previousFile = currentFile;
+		}
+	}
+	
+	@Test
+	public void getOutputFileToWrite_shouldFollowNamingConvention() throws IOException, ParseException {
+		String content = "A";
+		
+		// Construct 3 ComplexData objects and Obs to cover 3 cases - a given title, NULL title or a title without extension
+		ComplexData complexDataWithTitle = new ComplexData(FILENAME, content);
+		ComplexData complexDataWithNullTitle = new ComplexData(null, content);
+		ComplexData complexDataWithoutExtension = new ComplexData(FilenameUtils.removeExtension(FILENAME), content);
+		
+		Obs obsWithTitle = new Obs();
+		obsWithTitle.setComplexData(complexDataWithTitle);
+		
+		Obs obsWithNullTitle = new Obs();
+		obsWithNullTitle.setComplexData(complexDataWithNullTitle);
+		
+		Obs obsWithoutExtension = new Obs();
+		obsWithoutExtension.setComplexData(complexDataWithoutExtension);
+		
+		// Mocked methods
+		mockStatic(Context.class);
+		when(Context.getAdministrationService()).thenReturn(administrationService);
+		when(administrationService.getGlobalProperty(any())).thenReturn(complexObsTestFolder.newFolder().getAbsolutePath());
+		
+		AbstractHandler handler = new AbstractHandler();
+		
+		// Create files	
+		File titledFile = handler.getOutputFileToWrite(obsWithTitle);
```

Comment:
am I correct that you test 3 different cases?
if so please split them up in 3 tests. tests should be independent and if one case fails only one test should fail.
plus this might give you the opportunity to remove duplication in your tests by using a setup method or helper methods for setup/assertions...

Reply:
@teleivo You're right, I'll take a look.